### ISSUE TITLE
Is it worth it to sacrifice the visibility for modularization?

### ISSUE THREAD

**Issue by JerryWang0000** [nus-cs2103-AY2324S2/forum#562]
The question might be general. For example we need to put parts of codes into different packages. And then we need to modify the visibility of some functions which were supposed to be private helper functions. I know this side effect can be minimized through refactoring, but we still have such risks.

**Comment by damithc** [comment]
> For example we need to put parts of codes into different packages. And then we need to modify the visibility of some functions which were supposed to be private helper functions.

@JerryWang0000 Can you elaborate on this example a bit more? A class cannot access a private method of another class, irrespective of which package it is in. Perhaps you are referring to the [_extract class_ refactoring](https://refactoring.guru/extract-class) instead?

**Comment by JerryWang0000** [comment]
> > For example we need to put parts of codes into different packages. And then we need to modify the visibility of some functions which were supposed to be private helper functions.
> 
> @JerryWang0000 Can you elaborate on this example a bit more? A class cannot access a private method of another class, irrespective of which package it is in. Perhaps you are referring to the [_extract class_ refactoring](https://refactoring.guru/extract-class) instead?

Yes, exactly. My bad. I was thinking that, a class can have getter to retrieve one of its attributes. If we put it in another package, that means we might have to change getter to be "public", which could have been only seen within the package. Will that be considered undesired?

**Comment by J-wang-CSMA** [comment]
In the case of getters, in my experience they are usually just public methods written to access a field within a class while keeping the field itself private. So personally I don't think there's a tradeoff between visibility and modularization specifically for them. While there are some usecases for private getters I think their main utility is about letting the fields be accessed by other classes while hiding the fields.

As for other helper methods do you have a specific example in mind?

**Comment by tankh99** [comment]
I don't think there's any particular issue with making getters public as long as they are an attribute that the class itself owns.

**Comment by damithc** [comment]
To add, there are multiple reasons that can cause (or worsen) such a tradeoff. e.g.,
* splitting a class that should not be split
* avoidable coupling between classes
* not putting related classes in the same package
* leaky/faulty abstractions
* ...

Also, sometimes one _has_ to make a tradeoff, as it's not possible to have the best of everything.

It's hard to generalize this kind of thing. If there is a concrete example, we can dig deeper.
